### ISSUE TITLE
Set.equals fails to compare two identical Sets

### ISSUE THREAD

**Issue by patrox** [javaslang/javaslang#504]
This tests failes:

```java
@Test public void shouldMapNonEmpty() {
  final Set<Integer> expected = HashSet.<Integer>of(1, 2);
  final Set<Integer> actual =
    HashMap.<Integer, String>of(
      Map.Entry.of(1, "1"),
      Map.Entry.of(2, "2"))
    .map(entry -> entry.key);
  assertThat(actual).isEqualTo(expected);
}
```

with rather interesting message:

```java
java.lang.AssertionError: 
Expecting:
 <"HashSet(1, 2) (HashSet@8001)">
to be equal to:
 <"HashSet(1, 2) (HashSet@8001)">
but was not.
```

The minimal test to reproduce it is:

```java
@Test
public void shouldBeEqual() {
    assertTrue(HashSet.of(1).equals(HashSet.of(1)));
}
```

This seems to be caused by using `? (any type)` wildcard when casting `HashSet.equals` param to a HashSet class:

```
  final HashSet<?> that = (HashSet<?>) o;
```

The solution for this is issue is to change the type wildcard to `T` which is the type of elements for `this` `HashSet` instance, but this could result in `ClassCastException` when a `HashSet` of different type will be passed to `equals`.

**Comment by danieldietrich** [comment]
The underlying HashArrayMappedTrie may have stored the elements in an arbitrary order, so that the iterators do not return the elements in a specific order.

This lead to this implementation:

@SuppressWarnings("unchecked")
@Override
public boolean equals(Object o) {
    if (o == this) {
        return true;
    } else if (o instanceof HashSet) {
        final HashSet<?> that = (HashSet<?>) o;
        return this.length() == that.length() && ((HashSet<Object>) this).containsAll(that);
    } else {
        return false;
    }
}
We do not cast to <T> in order to prevent ClassCastExceptions.

**Comment by danieldietrich** [comment]
Fixed with #505
